File: api/src/main/java/api/security/model/JwtUserDetails.java
package api.security.model;

import api.entity.Role;
import api.entity.User;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;

import java.util.Collection;
import java.util.Set;

public class JwtUserDetails implements UserDetails {

    private Long id;
    private String username;
    private String password;
    private boolean isEnabled;
    private Set<Role> roles;
    private String refreshToken;

    public static JwtUserDetails create(User user){
        return new JwtUserDetails()
                .setAuthorities(user.getRoles())
                .setEnabled(user.isEnabled())
                .setId(user.getId())
                .setUsername(user.getUsername())
                .setPassword(user.getPassword())
                .setRefreshToken(user.getRefreshToken());
    }

    @Override
    public Collection<? extends GrantedAuthority> getAuthorities() {
        return roles;
    }

    @Override
    public String getPassword() {
        return password;
    }

    @Override
    public String getUsername() {
        return username;
    }

    @Override
    public boolean isAccountNonExpired() {
        return true;
    }

    @Override
    public boolean isAccountNonLocked() {
        return true;
    }

    @Override
    public boolean isCredentialsNonExpired() {
        return true;
    }

    @Override
    public boolean isEnabled() {
        return isEnabled;
    }

    public Long getId() {
        return id;
    }

    public JwtUserDetails setId(Long id) {
        this.id = id;
        return this;
    }

    public JwtUserDetails setUsername(String username) {
        this.username = username;
        return this;
    }

    public JwtUserDetails setPassword(String password) {
        this.password = password;
        return this;
    }

    public JwtUserDetails setEnabled(boolean enabled) {
        isEnabled = enabled;
        return this;
    }

    public JwtUserDetails setAuthorities(Set<Role> roles) {
        this.roles = roles;
        return this;
    }

    public String getRefreshToken() {
        return refreshToken;
    }

    public JwtUserDetails setRefreshToken(String refreshToken) {
        this.refreshToken = refreshToken;
        return this;
    }
}

File: api/src/main/java/api/security/model/JwtUserDetails.kt
package api.security.model

import api.entity.Role
import api.entity.User
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.userdetails.UserDetails

class JwtUserDetails : UserDetails {

    private var id: Long? = null
    private var username: String? = null
    private var password: String? = null
    private var isEnabled: Boolean = false
    private var roles: Set<Role> = mutableSetOf()
    private var refreshToken: String? = null

    override fun getAuthorities(): Collection<GrantedAuthority> = roles

    override fun getPassword(): String? = password

    override fun getUsername(): String? = username

    override fun isAccountNonExpired(): Boolean = true

    override fun isAccountNonLocked(): Boolean = true

    override fun isCredentialsNonExpired(): Boolean = true

    override fun isEnabled(): Boolean = isEnabled

    fun getId(): Long? = id

    fun setId(id: Long?): JwtUserDetails {
        this.id = id
        return this
    }

    fun setUsername(username: String?): JwtUserDetails {
        this.username = username
        return this
    }

    fun setPassword(password: String?): JwtUserDetails {
        this.password = password
        return this
    }

    fun setEnabled(enabled: Boolean): JwtUserDetails {
        isEnabled = enabled
        return this
    }

    fun setAuthorities(roles: Set<Role>): JwtUserDetails {
        this.roles = roles
        return this
    }

    fun getRefreshToken(): String? = refreshToken

    fun setRefreshToken(refreshToken: String?): JwtUserDetails {
        this.refreshToken = refreshToken
        return this
    }

    companion object {

        fun create(user: User): JwtUserDetails {
            return JwtUserDetails()
                    .setAuthorities(user.roles)
                    .setEnabled(user.isEnabled)
                    .setId(user.id)
                    .setUsername(user.username)
                    .setPassword(user.password)
                    .setRefreshToken(user.refreshToken)
        }
    }
}

fun JwtUserDetails.getUser(): User {
    return User().apply {
        id = getId()
        username = getUsername()
        isEnabled = isEnabled()
        roles = authorities.map { it as Role }.toMutableSet()
    }
}
